### PR TITLE
Improve Skylight resilience: token refresh, bounded paging, and full error logging

### PR DESCRIPTION
**Pull request by victorljn-earthranger**
## Summary

- **Token refresh**: Proactively check JWT expiry before using a cached token to avoid mid-request failures
- **Auth error handling**: On UNAUTHENTICATED/UNAUTHORIZED, clear the cached token so the next run re-authenticates — removed the ineffective backoff retry that reused the same stale transport headers
- **Bounded paging**: Page through events using Skylight's `meta.total` (`ceil(total / pageSize)`) instead of issuing an extra empty-page round-trip to detect the end of results
- **Resilient page failures**: Catch the `TransportError` base (not just `TransportQueryError`) at the page level, so **every** Skylight transport failure is handled — query errors, 500s (`TransportServerError`), protocol errors, etc. The failing page is skipped and the sync continues with events collected so far. If **page 1** fails (the page bound is not yet known) the AOI is abandoned, to avoid an unbounded loop
- **Full error logging**: An httpx response hook logs the complete request and response — URL, headers (with the `Authorization` token redacted), and both bodies — whenever Skylight returns a 4xx/5xx, so failures are fully diagnosable from the activity logs and shareable with the Skylight team. Successful responses are not logged (no noise, no large payload dumps)
- **None-response retry fix**: Replace the recursive `get_skylight_events` call on a `None` response with a bounded single retry that **rebuilds the request header + GraphQL client** before retrying, so the retry actually uses a fresh token (the previous behaviour reused the stale transport headers); gives up after the second consecutive `None`
- **Entry alerts**: Use the `start` point/time for `location` and `recorded_at` (not `end`); add `exit_date` and `duration_in_area` fields with a `"Pending"` fallback when the vessel hasn't exited yet
- **Config fix**: Raise `PullEventsBadConfigException` instead of returning the magic string `"error"` on bad event type mapping
- **URL cleanup**: Always use `DEFAULT_SKYLIGHT_API_URL`, drop the `integration.base_url` fallback
- **Dict mutation fix**: Use `{**default_transport_dict, 'headers': ...}` instead of mutating the original dict
- **UI**: Add the `SkylightEventType` enum with `uniqueItems=True` so the portal renders toggles instead of a free-text list. `dark_activity` intentionally omitted — deprecated
- **Dead code removal**: Remove the unused `SkylightEventTypes` enum and deprecated `dark_activity` entries from the mapping and `ERSkylightEventTypes`
- **Tests**: Add unit tests for the paging/error paths — last-page termination via `meta.total`, skip-failed-page-and-continue, page-1-failure break, 500 (`TransportServerError`) logging, and the `None`-response single-retry / give-up-after-two behaviour
- **Local dev**: `docker-compose` fixes — correct the Dockerfile path and add `redis` + pubsub emulator services and `env_file` so `docker compose up` works end-to-end
- **CI**: Add `--no-cache-dir` to the pip install step

## Scope

**Goal:** Improve resilience around token refresh and Skylight API failures, make every Skylight error fully diagnosable, and fix entry alert field mapping.

**Out of scope (tracked as follow-ups):**
- Attachment upload concurrency — pre-existing: the action times out at 9 min on large data windows due to sequential uploads; not introduced by this PR
- Skylight 500 root cause — pre-existing: full request/response is now logged (new in this PR) for investigation; resolution depends on the Skylight team
- Sync GQL client in an async handler — pre-existing: `gql_client.execute` blocks the event loop; replacing it with an async transport is a larger change
- `event_types` type annotation: `event_types` is typed as `List[SkylightEventType]` but holds `List[str]` after the `format_string_case` validator converts display labels to snake_case keys — the enum exists to constrain portal input, not as a runtime type. This is intentional: the enum formalises display labels that production config has used for months, with no behavioural change. Improving the type annotation is deferred to a follow-up PR

**Quality bar:** No regressions. The gaps above existed before this PR; this PR does not make them worse.

## Test plan

- [x] All unit tests passing (`pytest app/actions/tests/`)
- [x] Pull events returns events for a configured AOI (verified locally against staging)
- [x] Paging terminates via `meta.total` without an extra empty-page request
- [ ] Auth action returns `{"valid_credentials": true}` locally
- [ ] Entry alert events have correct location (start point) and `exit_date`/`duration_in_area` fields
- [ ] A Skylight error on one page surfaces in activity logs with the full request + response and does not discard events from previous pages
- [ ] Portal shows the toggle UI for the event types field

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: app/actions/client.py [hunk 1]
```diff
@@ -283,20 +294,13 @@ async def get_skylight_events(integration, config_data, auth):
...
-    # get Token
-    headers = await build_request_header(integration, auth, gql_client)
-
-    # Add 'header' to GraphQL client
-    transport_dict_with_header = default_transport_dict
-    transport_dict_with_header['headers'] = headers.dict()
-    gql_client = build_graphql_client(transport_dict_with_header)
+    auth_client = build_graphql_client(default_transport_dict)
+    headers = await build_request_header(integration, auth, auth_client)
+    gql_client = build_graphql_client({**default_transport_dict, 'headers': headers.dict()})
```
[the remark this answers is not in the document]
Comment (victorlujanearthranger) [reply app/actions/client.py]: Fixed — the auth path in `handlers.py` now also uses `DEFAULT_SKYLIGHT_API_URL` directly, consistent with `get_skylight_events`. PR description updated accordingly.

File: app/actions/client.py [hunk 2]
```diff
@@ -169,18 +164,34 @@ def build_graphql_client(transport_dict):
...
         fetch_schema_from_transport=False
     )
     return gql_client
 
 
+def _is_token_expired(access_token: str) -> bool:
+    """Best-effort JWT exp check. Returns True on any parse failure (fail-safe)."""
+    try:
+        payload = access_token.split(".")[1]
+        padded = payload + "=" * ((-len(payload)) % 4)
```
[the remark this answers is not in the document]
Comment (victorlujanearthranger) [reply app/actions/client.py:224]: Added unit tests covering: valid cached token reuse (no re-auth), refresh when token is within the 60s skew window, refresh when token is expired, and fresh fetch when there is no cached state. Also added isolated `_is_token_expired` tests for the unparsable and empty token cases.

File: app/actions/handlers.py [hunk 3]
```diff
@@ -95,8 +95,37 @@ def transform(config, data: dict) -> dict:
...
+        is_entry_alert = (
+            _skylight_type == "aoi_visit"
+            or (isinstance(_skylight_type, list) and "aoi_visit" in _skylight_type)
+        )
+
+        if is_entry_alert:
+            event_time_and_location = data.get('start')
+            if not event_time_and_location:
+                logger.warning(f"Entry alert '{data.get('event_id')}' has no start point, skipping.")
+                return {}
```
[the remark this answers is not in the document]
Comment (victorlujanearthranger) [reply app/actions/handlers.py:108]: Added three focused `transform` tests for the `aoi_visit` path: (1) start+end present — verifies `recorded_at` and `location` come from `start`, and that `exit_date`/`duration_in_area` are computed correctly from the delta; (2) start-only — verifies `"Pending"` fallback for both fields; (3) missing start — verifies `transform` returns `{}` and skips the event.

File: app/actions/configurations.py [hunk 4]
```diff
@@ -29,9 +44,10 @@ class PullEventsConfig(PullActionConfiguration):
...
         description='IDs of the desired areas.',
     )
-    event_types: List[str] = Field(
+    event_types: List[SkylightEventType] = Field(
         title='Event Types to Fetch',
         description='The list of EventTypes the integration will use.',
+        uniqueItems=True,
     )
     pageSize: int = Field(
         1000,
```
[the remark this answers is not in the document]
Comment (victorlujanearthranger) [reply app/actions/configurations.py:53]: Good catch on the surface, but this behaviour predates this PR. The portal JSON schema was manually edited months ago to use these exact display labels (`"Dark Rendezvous"`, `"Fishing"`, etc.) as enum values — meaning production configs have always stored display names, not snake_case keys. The `format_string_case` validator was already in place to convert them. The new `SkylightEventType` enum simply formalises in code what was already hardcoded in the manually-edited JSON schema — no behaviour change. Verified: all 6 display labels from the production portal schema round-trip correctly through the validator and resolve to a valid `DEFAULT_EVENT_MAPPING` entry. Added a comment in the validator to document this history so it's clear to future reviewers.

File: app/actions/client.py [hunk 5]
```diff
@@ -241,40 +252,40 @@ async def get_authentication_token(integration, auth, gql_client):
...
-        events_mapping = integration.additional or DEFAULT_EVENT_MAPPING
-        parsed_obj = EventType.parse_obj(events_mapping.get(event_type, {}))
+        return EventType.parse_obj(events_mapping.get(event_type, {}))
     except pydantic.ValidationError as e:
         message = f'Failed to map "{event_type}". Either is invalid or unsupported. {e}'
         logger.warning(message)
-        return "error"
-    else:
-        return parsed_obj
+        raise PullEventsBadConfigException(message)
```
[the remark this answers is not in the document]
Comment (victorlujanearthranger) [reply app/actions/client.py:308]: Same context as the `configurations.py` comment: by the time `map_event_type` is called, `format_string_case` has already converted the display label to a snake_case key (e.g. `"Dark Rendezvous"` → `"dark_rendezvous"`), which matches `DEFAULT_EVENT_MAPPING` exactly. This was verified empirically against the production portal config. The lookup does not receive raw enum values or display names — it always receives the already-converted string. Added a comment in the validator explaining this flow.

File: app/actions/client.py [hunk 6]
```diff
@@ -416,22 +419,39 @@ async def get_skylight_events(integration, config_data, auth):
...
                     logger.info(f'"getRendedzvousExternal" query returned None, retrying with a new token...')
-                    await state_manager.delete_state(
-                        str(integration.id),
-                        "pull_events",
-                        auth.username
-                    )
-                    return await get_skylight_events(integration, config_data, auth)
+                    await state_manager.delete_state(str(integration.id), "pull_events", auth.username)
+                    none_retries += 1
+                    continue
```
[the remark this answers is not in the document]
Comment (victorlujanearthranger) [reply app/actions/client.py:513]: Good catch, thank you. The `gql_client` is built once before the paging loop with the original auth headers baked into the transport; deleting the cached token state and `continue`-ing left the same stale `Authorization` header in place, so the retry would hit Skylight with the same credentials and fail identically.

**Before:**
```python
await state_manager.delete_state(str(integration.id), "pull_events", auth.username)
none_retries += 1
continue
```

**After:**
```python
await state_manager.delete_state(str(integration.id), "pull_events", auth.username)
headers = await build_request_header(integration, auth, auth_client)
gql_client = build_graphql_client({**default_transport_dict, 'headers': headers.dict()})
none_retries += 1
continue
```

File: app/actions/client.py [hunk 7]
```diff
@@ -416,22 +419,39 @@ async def get_skylight_events(integration, config_data, auth):
...
+                        f'"getRendedzvousExternal" page {page_num} failed for AOI "{aoi}": {te}. '
+                        f'Request params: {params}. '
+                        f'Keeping {len(response_list)} events collected so far.',
+                        extra={
+                            "integration_id": str(integration.id),
+                            "aoi": aoi,
+                            "attention_needed": True,
+                        }
+                    )
+                    break
```
[the remark this answers is not in the document]
Comment (victorlujanearthranger) [reply app/actions/client.py]: Thank you — you're right on both counts. The page-level `TransportQueryError` now skips just the failing page and continues, and the loop is bounded by Skylight's `meta.total` (we compute `total_pages` from the first page and stop there), so there's no infinite-loop risk even if every page fails.

**Before:**
```python
while has_data:
    try:
        response = await execute_gql_query(...)
    except TransportQueryError as te:
        logger.error(...)
        break  # stops paging the whole AOI
    ...
    if not events_response:
        has_data = False  # needs an extra empty-page round-trip to terminate
    response_list.extend(events_response)
    page_num += 1
```

**After:**
```python
total_pages = None
while total_pages is None or page_num <= total_pages:
    try:
        response = await execute_gql_query(...)
    except TransportQueryError as te:
        logger.error(...)
        page_num += 1
        continue  # skip the bad page, keep going
    ...
    meta = response['events']['meta']
    if total_pages is None:
        total = meta['total'] or 0
        total_pages = (total + meta['pageSize'] - 1) // meta['pageSize']
    response_list.extend(events_response)
    page_num += 1
```

File: app/actions/client.py [hunk 8]
```diff
@@ -416,22 +419,39 @@ async def get_skylight_events(integration, config_data, auth):
...
                 logger.info(f'Sending "getRendedzvousExternal" query request. Params: "{params}"...')
 
-                response = await execute_gql_query(gql_client, query, params, integration, auth)
+                try:
+                    response = await execute_gql_query(gql_client, query, params, integration, auth)
+                except TransportQueryError as te:
+                    # Log as error with attention_needed so this surfaces in
+                    # activity logs. We break rather than raise so a partial
+                    # page failure doesn't abort the entire AOI sync.
+                    logger.error(
```
[the remark this answers is not in the document]
Comment (victorlujanearthranger) [reply app/actions/client.py:475]: Thank you — added focused coverage for the new paging/error paths in `get_skylight_events` (4 tests in `app/actions/tests/test_skylight.py`):

- `test_get_skylight_events_stops_at_last_page_via_meta` — loop terminates via `meta.total` page math, no wasted empty-page round-trip.
- `test_get_skylight_events_skips_failed_page_and_continues` — a `TransportQueryError` on page 2 of 3 skips only that page; pages 1 and 3 are still collected.
- `test_get_skylight_events_retries_once_on_none_with_new_client` — a `None` items response clears the token, rebuilds the GraphQL client, and retries the page once.
- `test_get_skylight_events_gives_up_after_two_none_responses` — bails (no infinite retry) after the second consecutive `None`.

All four pass; the existing suite is otherwise green.

File: app/actions/client.py [hunk 9]
```diff
@@ -416,25 +456,58 @@ async def get_skylight_events(integration, config_data, auth):
...
 
-                logger.info(f'"getRendedzvousExternal" query returned: "{events_response}"...')
-
                 if events_response is None:
-                    # Currently, if we receive None in the response, there was an unknown error in Skylight API
-                    # Will delete token and try again
+                    if none_retries >= 1:
+                        logger.error(
+                            f'"getRendedzvousExternal" returned None twice for AOI "{aoi}", giving up.',
+                            extra={"integration_id": str(integration.id), "aoi": aoi, "attention_needed": True}
```
[the remark this answers is not in the document]
Comment (victorlujanearthranger) [reply app/actions/client.py:505]: Good catch, thank you. `none_retries` was per-AOI, so a single earlier `None` (even one we recovered from) armed the counter and would abort on the next `None` much later — "two Nones per AOI" rather than "two consecutive". Now reset to 0 after any successful page.

**Before:**
```python
meta = response['events']['meta']
if total_pages is None:
    ...
response_list.extend(events_response)
page_num += 1
```

**After:**
```python
# Successful page: reset the None counter so only *consecutive*
# None responses (not Nones scattered across the AOI) trigger the abort.
none_retries = 0

meta = response['events']['meta']
if total_pages is None:
    ...
response_list.extend(events_response)
page_num += 1
```

Added `test_get_skylight_events_resets_none_counter_after_success` which fails if the reset is removed (None on page 1 recovered, None on page 3 recovered, all events collected).

File: app/actions/tests/test_skylight.py [hunk 10]
```diff
@@ -97,6 +109,227 @@ async def test_execute_gql_query_does_not_retry_if_not_unauthorized_code(
...
+            _page([{"event_id": "e1"}, {"event_id": "e2"}], total=4, page_num=1),
+            TransportServerError("500 Internal Server Error"),
+        ],
+    )
+
+    events, _ = await get_skylight_events(integration, _PullCfg(), auth)
+
+    assert exec_mock.call_count == 2
+    assert [e["event_id"] for e in events["aoi1"]] == ["e1", "e2"]
+
```
[the remark this answers is not in the document]
Comment (victorlujanearthranger) [reply app/actions/tests/test_skylight.py:246]: Thank you — you're right, the test name claimed "logs" but only asserted the skip. Added an assertion that the page-level `logger.error` fired with `attention_needed` so the test verifies the logging behavior it describes.

**After:**
```python
log = mocker.patch("app.actions.client.logger")

events, _ = await get_skylight_events(integration, _PullCfg(), auth)

assert exec_mock.call_count == 2
assert [e["event_id"] for e in events["aoi1"]] == ["e1", "e2"]
assert any(
    "failed for AOI" in str(call) and call.kwargs.get("extra", {}).get("attention_needed")
    for call in log.error.call_args_list
)
```

Also added dedicated tests for the error-response hook itself (full request/response logged on 4xx/5xx, bearer token redacted, silent on success).

File: app/actions/handlers.py [hunk 11]
```diff
@@ -95,8 +95,37 @@ def transform(config, data: dict) -> dict:
...
+                    full_event_details["duration_in_area"] = "Pending"
+            else:
+                full_event_details["exit_date"] = "Pending"
+                full_event_details["duration_in_area"] = "Pending"
+        else:
+            event_time_and_location = data.get('end') or data.get('start')
+
+        if not event_time_and_location:
+            logger.warning(f"Event '{data.get('event_id')}' has no start or end point, skipping.")
+            return {}
```
Comment (chrisdoehring) [inline app/actions/handlers.py:128]: **Fix before merge:** these new `{}` returns leak into the Gundi batch at the call site.

`transform()` now returns `{}` in more cases (entry alert with no `start`, line 108; this general guard, line 128). But `action_process_events_per_aoi` (~line 277) builds `[transform(...) for event in ...]` and then checks `if transformed_data:` — a list containing `{}` is still truthy, so the empty dicts survive into `generate_batches(...)` / `send_events_to_gundi`.

Note `patch_events` (~line 388) already filters per-event with `if transformed_data:`, so the two call sites are inconsistent. Suggest filtering falsy results before sorting at ~line 277:

```python
transformed_data = sorted(
    [t for event in action_config.events if (t := transform(action_config.updated_config_data, event))],
    key=lambda x: x.get("recorded_at") or datetime.datetime.min, reverse=True,
)
```

Pre-existing (lines 56/60 already produced `{}`), but this PR widens the exposure, so worth closing here.

File: app/actions/client.py [hunk 12]
```diff
@@ -416,25 +466,63 @@ async def get_skylight_events(integration, config_data, auth):
...
+                    headers = await build_request_header(integration, auth, auth_client)
+                    gql_client = build_events_client(default_transport_dict, headers)
+                    none_retries += 1
+                    continue
+
+                # Successful page: reset the None counter so only *consecutive*
+                # None responses (not Nones scattered across the AOI) trigger the abort.
+                none_retries = 0
+
+                meta = response['events']['meta']
```
Comment (chrisdoehring) [inline app/actions/client.py]: **Fix or file a follow-up ticket:** `meta` is assumed non-null.

If Skylight ever returns `meta: null`, `meta.get('total')` raises `AttributeError`, which falls through to the broad `except Exception` at ~line 551 → the entire AOI is dropped, **including the events already in `response_list`** (`events.update({aoi: response_list})` at ~line 529 is never reached). That partially defeats the "keep events collected so far" goal of this PR.

Cheap guard:
```python
meta = response['events'].get('meta') or {}
```

If you'd rather not expand scope here, fine to defer — please file a Jira follow-up so it isn't lost.

**Review by chrisdoehring — COMMENTED** [review]
Reviewed — solid, well-scoped resilience work. Worth calling out: **the in-diff documentation is excellent.** The comments explaining the SkylightEventType-enum / `format_string_case` validator / `DEFAULT_EVENT_MAPPING` lockstep, the page-1-failure reasoning (why we abandon the AOI when `total_pages` isn't yet known), the token-skew rationale, and the sync-vs-async note on the response hook all make the intent obvious and the change easy to maintain. The PR description's explicit in/out-of-scope split is also great. Two inline items below — one to fix before merge, one to fix or ticket.

File: app/actions/handlers.py (hunk 11 above)
Comment (victorlujanearthranger) [reply app/actions/handlers.py:128]: Good catch, thank you. `transform()` now returns `{}` in more cases, and since a list like `[{}]` is truthy the `if transformed_data:` check didn't stop empty dicts from flowing into `generate_batches`/`send_events_to_gundi`. Now filtered at the source (and consistent with the per-event filtering in `patch_events`).

**Before:**
```python
transformed_data = sorted(
    [transform(action_config.updated_config_data, event) for event in action_config.events],
    key=lambda x: x.get("recorded_at") or datetime.datetime.min, reverse=True
)
```

**After:**
```python
transformed_data = sorted(
    [
        transformed
        for event in action_config.events
        if (transformed := transform(action_config.updated_config_data, event))
    ],
    key=lambda x: x.get("recorded_at") or datetime.datetime.min, reverse=True
)
```

Added `test_action_process_events_per_aoi_filters_empty_transforms` asserting `{}` never enters the batch.

File: app/actions/client.py (hunk 12 above)
Comment (victorlujanearthranger) [reply app/actions/client.py]: Thank you — you're right, and it directly undercut the "keep events collected so far" goal. If Skylight returned `meta: null`, `meta.get('total')` would raise `AttributeError`, hit the broad `except Exception`, and drop the AOI before `events.update({aoi: response_list})` ran — discarding everything already collected. Added the cheap guard.

**Before:**
```python
meta = response['events']['meta']
if total_pages is None:
    total = meta.get('total') or 0
```

**After:**
```python
# Guard against a null/absent meta: without this, meta.get() would raise
# AttributeError, fall through to the broad except below, and drop the AOI
# *including events already collected*.
meta = response['events'].get('meta') or {}
if total_pages is None:
    total = meta.get('total') or 0
```

Added `test_get_skylight_events_handles_null_meta` verifying a null meta keeps collected events instead of crashing the AOI.

**Review by chrisdoehring — APPROVED** [review]
Thanks Victor.
